File: 프로그래머스/level4/호텔 방 배정.cpp
#include <string>
#include <map>
#include <vector>
#include <iostream>

using namespace std;

map<long long, long long> mp;

long long Find(long long now){
    if(mp.find(now)==mp.end()){
        mp[now] = now + 1;
        return now+1;
    }
    return mp[now] = Find(mp[now]);
}

vector<long long> solution(long long k, vector<long long> room_number) {
    vector<long long> answer;
    for(int i=0; i<room_number.size(); i++){
        long long want = room_number[i];
        long long f = Find(want);
        answer.push_back(f-1);
    }
    return answer;
}

File: 프로그래머스/level4/호텔 방 배정.js
let obj = new Map();

let Find = (want) => {
    if(obj.get(want)===undefined){
        obj.set(want, want+1);
        return want;
    }
    obj.set(want, Find(obj.get(want)));
    return obj.get(want);
}

function solution(k, room_number) {
    var answer = [];
    room_number.forEach(num=>{
        let result = Find(num);
        answer.push(result);
    });
    return answer;
}